### PR TITLE
Don't push out the outputs of an 'as_while' Scan

### PR DESCRIPTION
**Pull request by carriepl**
this is preemtive fix. It could cause an crash or bad value. as_while isn't much used.

### CODE REVIEW

File: theano/scan_module/scan_opt.py [hunk 1]
```diff
@@ -612,8 +612,11 @@ def add_requirements(self, fgraph):
         fgraph.attach_feature(gof.toolbox.ReplaceValidate())
 
     def apply(self, fgraph):
+        # Don't perform the optimization on as_while scans since it's
+        # impossible to know for how many steps it will run.
         nodelist = [x for x in fgraph.toposort()
-                    if isinstance(x.op, scan_op.Scan)]
+                    if (isinstance(x.op, scan_op.Scan) and
+                        not x.op.as_while]
```
Comment (nouiz) [inline theano/scan_module/scan_opt.py]: This need a test. Seem ok for now, but I don't agree we can never push output of scan in while mode. After the scan, we know how many iteration was done. But we shouldn't push outputs used for by the condition. But we can let this for later.

Comment (nouiz) [inline theano/scan_module/scan_opt.py]: So just update the comment for this if you agree.

**Comment by nouiz** [comment]
I think this is good, but @abergeron can you also review it?

**Comment by lamblin** [comment]
OK for me as well.

**Comment by carriepl** [comment]
Fixed an issue with running the new test in floatX=float32

**Comment by nouiz** [comment]
as travis is overloaded, I killed it on the last commit. I'm sure it is good. I merge now.

**Pull request merged**